examples/src/test/java/io/grpc/examples/header/HeaderServerInterceptorTest.java

```diff
@@ -0,0 +1,138 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.examples.header;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Server;
+import io.grpc.ServerInterceptors;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterBlockingStub;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterImplBase;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.StreamObserver;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+
+import java.io.IOException;
+
+/**
+ * Unit tests for {@link HeaderClientInterceptor}.
+ * For demonstrating how to write gRPC unit test only.
+ * Not intended to provide a high code coverage or to test every major usecase.
+ *
+ * <p>For basic unit test examples see {@link io.grpc.examples.helloworld.HelloWorldClientTest} and
+ * {@link io.grpc.examples.helloworld.HelloWorldServerTest}.
+ */
+@RunWith(JUnit4.class)
+public class HeaderServerInterceptorTest {
+  private Server fakeServer;
+  private ManagedChannel inProcessChannel;
+
+  @Before
+  public void setUp() throws IOException {
+    String uniqueServerName = "fake server for " + getClass();
+    GreeterImplBase greeterImplBase =
+        new GreeterImplBase() {
+          @Override
+          public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
+            responseObserver.onNext(HelloReply.getDefaultInstance());
+            responseObserver.onCompleted();
+          }
+        };
+    fakeServer = InProcessServerBuilder.forName(uniqueServerName)
+        .addService(ServerInterceptors.intercept(greeterImplBase, new HeaderServerInterceptor()))
+        .directExecutor()
+        .build()
+        .start();
+    inProcessChannel = InProcessChannelBuilder.forName(uniqueServerName).directExecutor().build();
+  }
+
+  @After
+  public void tearDown() {
+    inProcessChannel.shutdownNow();
+    fakeServer.shutdownNow();
+  }
+
+  @Test
+  public void serverHeaderDeliveredToClient() {
+    class SpyingClientInterceptor implements ClientInterceptor {
+      ClientCall.Listener<?> spyListener;
```

Comment:
The correct way to do this is with NoopClientCallListener and not spying on the implementation